### PR TITLE
Fix errors and openai dict

### LINKED ISSUE #168: [BUG] Dictionary Type not working with OpenAI Extractor
Not sure if a bug or a feature request, as it might be related to OpenAI. Extracting to a Pydantic model that contains a dictionary throws validation errors when Pydantic tries to convert the returned string to a model.

**To Reproduce**
Copy and Paste these code blocks into a notebook.
Fill in your API key on this notebook, run the cells.

```
from typing import Dict, List, Type
from pydantic import BaseModel
from mirascope.openai import OpenAICall, OpenAICallParams, OpenAIExtractor
import json
import os

os.environ["OPENAI_API_KEY"] = "<Api-Key>"

class FoodExamples(BaseModel):
    foods_and_examples : Dict[str, List[str]]

class FoodExamplesExtractor(OpenAIExtractor[FoodExamples]):
    extract_schema: Type[FoodExamples] = FoodExamples
    prompt_template = """
    Extract the food types and their specific examples from the following json string:
    {json_string}
    """
    json_string : str 
    call_params = OpenAICallParams(model = "gpt-4")

simple_dict = {
    "cheese" : [
        "parmesean",
        "pecorino",
        "mozzarella"
    ],
    "bread" : [
        "sourdough",
        "whole wheat",
        "rye"
    ]
}

food_examples = FoodExamplesExtractor(json_string=json.dumps(simple_dict)).extract()
assert isinstance(food_examples, FoodExamples)
print(food_examples)
```

### PR DESCRIPTION
**Pull request by willbakst**
Closes #168 

- Raise RuntimeError when stop reason isn't function calls when using `tools` and `tool` convenience functions.
- Enable using `response_format={"type": "json_object"}` when using tools and extraction.
- Update Anthropic system prompt when using extract to force it to use a tool correctly.

### CODE REVIEW

File: mirascope/anthropic/types.py [hunk 1]
```diff
@@ -90,12 +91,19 @@ class BookRecommender(AnthropicCall):
...
-            not self.tool_types
-            or self.response.stop_reason != "stop_sequence"
+            self.response.stop_reason != "stop_sequence"
             or self.response.stop_sequence != "</function_calls>"
         ):
-            return None
+            raise RuntimeError(
+                "Generation stopped before `</function_calls>` stop sequence. "
+                "This is likely due to a limit on output tokens that is too low. "
+                "Note that this could also indicate no tool is beind called, so we "
```
Comment (brenkao) [inline mirascope/anthropic/types.py:104]: would be nice to have a copy and pastable command that people can use to check. something like call.response.stop_reason or something.

Comment (willbakst) [reply mirascope/anthropic/types.py:104]: I should just include it in the text. good call

File: mirascope/openai/calls.py [hunk 2]
```diff
@@ -185,3 +203,31 @@ async def stream_async(
...
+        if self.call_params.response_format == {"type": "json_object"} and tool_types:
+            kwargs.pop("tools")
+        return kwargs, tool_types
+
+    def _maybe_update_messages(
+        self,
+        messages: list[ChatCompletionMessageParam],
+        tool_types: Optional[list[type[OpenAITool]]],
+    ) -> list[ChatCompletionMessageParam]:
+        if self.call_params.response_format == {"type": "json_object"} and tool_types:
```
Comment (brenkao) [inline mirascope/openai/calls.py:227]: technically json is the format so we can omit the object part

Comment (willbakst) [reply mirascope/openai/calls.py:227]: https://platform.openai.com/docs/guides/text-generation/json-mode

See line 6 of the code block in this section

File: mirascope/openai/calls.py [hunk 3]
```diff
@@ -185,3 +203,31 @@ async def stream_async(
...
+        dict[str, Any],
+        Optional[list[Type[OpenAITool]]],
+    ]:
+        """Overrides the `BaseCall._setup` for Anthropic specific setup."""
+        kwargs, tool_types = self._setup(kwargs, OpenAITool)
+        if self.call_params.response_format == {"type": "json_object"} and tool_types:
+            kwargs.pop("tools")
+        return kwargs, tool_types
+
+    def _maybe_update_messages(
```
Comment (brenkao) [inline mirascope/openai/calls.py:222]: change function name to be `_update_message_if_json` so its more clear than maybe

**Review by brenkao — APPROVED** [review]

File: mirascope/openai/calls.py (hunk 2 above)
Comment (willbakst) [reply mirascope/openai/calls.py]: Used OpenAI's ResponseFormat object

File: mirascope/openai/calls.py (hunk 3 above)
Comment (willbakst) [reply mirascope/openai/calls.py]: Done.

**Pull request merged**